### PR TITLE
Revision of DateTime.Parse overloads

### PR DESCRIPTION
**Pull request by rpetrusha**
# Revision of DateTime.Parse overloads

[Internal review link - Parse](https://review.docs.microsoft.com/en-us/dotnet/api/system.datetime.parse?view=netframework-4.7.1&branch=pr-en-us-4455)
[Internal review link - ParseExact](https://review.docs.microsoft.com/en-us/dotnet/api/system.datetime.parseexact?view=netframework-4.7.1&branch=pr-en-us-4455)

Edit by @billwagner: Add internal review links.

### CODE REVIEW

File: samples/snippets/csharp/VS_Snippets_CLR_System/system.DateTime.Parse/cs/Parse6.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
+using System;
+
+public class Example
+{
+   public static void Main()
+   {
+      ValueTuple<string,string>[] dateInfo = { ("08/18/2018 07:22:16", "String with a date and time component"),
```
Comment (BillWagner) [inline samples/snippets/csharp/VS_Snippets_CLR_System/system.DateTime.Parse/cs/Parse6.cs]: I'd write the declaration like this:

```csharp
(string dateAsString, string Description)[] dateInfo = {}
```

Then, you can use `dateAsString` and `Description` instead of `Item1` and `Item2`

File: samples/snippets/visualbasic/VS_Snippets_CLR_System/system.DateTime.Parse/vb/Parse2.vb [hunk 2]
```diff
@@ -4,15 +4,15 @@ Option Strict On
...
+                                     "2008-05-01 7:34:42Z", 
                                      "Thu, 01 May 2008 07:34:42 GMT"}
       
-      For Each dateString As String In dateStrings
+      For Each dateString In dateStrings
          Dim convertedDate As Date = Date.Parse(dateString)
-         Console.WriteLine("Converted {0} to {1} time {2}", _
-                           dateString, _
-                           convertedDate.Kind.ToString(), _
+         Console.WriteLine("Converted {0} to {1} time {2}",
```
Comment (BillWagner) [inline samples/snippets/visualbasic/VS_Snippets_CLR_System/system.DateTime.Parse/vb/Parse2.vb]: I'd use the interpolated string syntax here. It would be more clear for the reader.

File: samples/snippets/visualbasic/VS_Snippets_CLR_System/system.DateTime.Parse/vb/Parse5.vb [hunk 3]
```diff
@@ -1,45 +1,42 @@
...
+      Dim formattedDates() = { "2008-09-15T09:30:41.7752486-07:00", 
+                                 "2008-09-15T09:30:41.7752486Z",  
+                                 "2008-09-15T09:30:41.7752486",  
+                                 "2008-09-15T09:30:41.7752486-04:00", 
+                                 "Mon, 15 Sep 2008 09:30:41 GMT" }
+      For Each formattedDate In formattedDates
+         Console.WriteLine(formattedDate)
+         Dim roundtripDate = DateTime.Parse(formattedDate, Nothing,  
+                                                 DateTimeStyles.RoundtripKind)                        
+         Console.WriteLine("   With RoundtripKind flag: {0} {1} time.",
```
Comment (BillWagner) [inline samples/snippets/visualbasic/VS_Snippets_CLR_System/system.DateTime.Parse/vb/Parse5.vb]: another good use for interpolated strings.

File: samples/snippets/visualbasic/VS_Snippets_CLR_System/system.DateTime.Parse/vb/Parse6.vb [hunk 4]
```diff
@@ -0,0 +1,35 @@
+
+Public Module Strings
+   Public Sub Main()
+      Dim dateInfo() As ValueTuple(Of String, String) = { ("08/18/2018 07:22:16", "String with a date and time component"),
```
Comment (BillWagner) [inline samples/snippets/visualbasic/VS_Snippets_CLR_System/system.DateTime.Parse/vb/Parse6.vb]: Similar to the C# example, I'd declare this as:

```visualbasic
Dim dateInfo() As (DateString As String, Description As String)
```

File: xml/System/DateTime.xml [hunk 5]
```diff
@@ -4236,130 +4236,79 @@ juillet 2009
...
 |Parse a date and time string and perform a conversion to UTC or local time.|<xref:System.DateTime.Parse%28System.String%2CSystem.IFormatProvider%2CSystem.Globalization.DateTimeStyles%29> overload|  
 |Parse a date and time string without handling exceptions.|<xref:System.DateTime.TryParse%2A?displayProperty=nameWithType> method|  
 |Restore (round-trip) a date and time value created by a formatting operation.|Pass the "o" or "r" standard format string to the <xref:System.DateTime.ToString%28System.String%29> method, and call the <xref:System.DateTime.Parse%28System.String%2CSystem.IFormatProvider%2CSystem.Globalization.DateTimeStyles%29> overload with <xref:System.Globalization.DateTimeStyles.RoundtripKind?displayProperty=nameWithType>|  
 |Parse a date and time string in a fixed format across machine (and possibly cultural) boundaries.|<xref:System.DateTime.ParseExact%2A?displayProperty=nameWithType> or <xref:System.DateTime.TryParseExact%2A?displayProperty=nameWithType> method|  
   
 <a name="StringToParse"></a>   
-## The string to parse  
+### The string to parse
```
Comment (BillWagner) [inline xml/System/DateTime.xml]: The variable `s` has not been introduced here. I'd reword that sentence to "the input string"

File: xml/System/DateTime.xml [hunk 6]
```diff
@@ -4236,130 +4236,79 @@ juillet 2009
...
   
 > [!IMPORTANT]
->  Because the string representation of a date and time must conform to a recognized pattern, you should always use exception handling when calling the <xref:System.DateTime.Parse%2A> method to parse user input. If you don't want to handle exceptions, you can call the <xref:System.DateTime.TryParse%2A?displayProperty=nameWithType> method to parse the date and time string; this method returns a value that indicates whether the parse operation succeeded.  
+>  Because the string representation of a date and time must conform to a recognized pattern, you should always use exception handling when calling the <xref:System.DateTime.Parse%2A> method to parse user input. If you don't want to handle exceptions, you can call the <xref:System.DateTime.TryParse%2A?displayProperty=nameWithType> method to parse the date and time string; this method returns a <xref:System.Boolean> value that indicates whether the parse operation succeeded.
```
Comment (BillWagner) [inline xml/System/DateTime.xml]: I'd word this differently: `Parse` would be better when the operation is expected to succeed because the input source is trusted. `TryParse` is preferred when parse failures are likely, or you have reasonable defaults. The distinction is more about *how* you want to handle errors than *if* you want to handle them.

File: xml/System/DateTime.xml [hunk 7]
```diff
@@ -4236,130 +4236,79 @@ juillet 2009
...
+     "03/01/2009 05:42:00 -5:00"  
+
+The following example parses strings in each of these formats by using the formatting conventions of the current culture, which in this case is the en-US culture:
+
+[!code-csharp-interactive[Default parsing](~/samples/snippets/csharp/VS_Snippets_CLR_System/system.DateTime.Parse/cs/Parse6.cs)]
+[!code-vb[Default parsing](~/samples/snippets/visualbasic/VS_Snippets_CLR_System/system.DateTime.Parse/vb/Parse6.vb)]  
   
- If `s` is the string representation of a leap day in a leap year in the calendar used by the parsing method (see [Parsing and cultural conventions](#Culture)), the <xref:System.DateTime.Parse%2A> method parses `s` successfully. If `s` is the string representation of a leap day in a non-leap year, the method throws a <xref:System.FormatException> exception.  
+ If `s` is the string representation of a leap day in a leap year in the calendar used by the parsing method (see [Parsing and cultural conventions](#Culture)), the <xref:System.DateTime.Parse%2A> method parses `s` successfully. If `s` is the string representation of a leap day in a non-leap year, the method throws a <xref:System.FormatException>.
```
Comment (BillWagner) [inline xml/System/DateTime.xml]: same comment on variable `s`: it hasn't been introduced. Since it is used in the samples below, it may be best to introduce it here.

File: xml/System/DateTime.xml [hunk 8]
```diff
@@ -4571,18 +4545,31 @@ An object that is equivalent to the date and time contained in `s`.
...
+        <summary>Converts the string representation of a date and time to its <see cref="T:System.DateTime" /> equivalent by using culture-specific format information and a formatting style.</summary>
         <returns>An object that is equivalent to the date and time contained in <paramref name="s" />, as specified by <paramref name="provider" /> and <paramref name="styles" />.</returns>
         <remarks>
           <format type="text/markdown"><![CDATA[  
   
 ## Remarks  
+
+This method overload converts the date and time in `s` and sets the <xref:System.DateTime.Kind%2A> property of the returned <xref:System.DateTime> value as follows:
+
+|If |Time zone conversion|Kind property|
```
Comment (BillWagner) [inline xml/System/DateTime.xml]: For some reason, this table is not rendering properly. I think the 2nd line only has two columns, instead of three.

**Review by BillWagner — CHANGES_REQUESTED** [review]
I really like what you've done here @rpetrusha Overall, this is a great improvement. 

In particular, I really like the table "which method to call". That's clear and actionable.

I left a few comments in the markdown and the source. to address.  
Also, we should consider how we want to handle the sample code. I had made a sample project when I updated the page for the `DateTime` struct: https://github.com/dotnet/docs/samples/snippets/csharp/System.DateTime and https://github.com/dotnet/docs/samples/snippets/visualbasic/System.DdateTime.

It occurs to me that I put those in the wrong directory. It may make sense to combine the samples you did here, and those, and put them all in a better location. I think that's a separate PR though. If you agree, let's open an issue on that for later.

File: xml/System/DateTime.xml [hunk 9]
```diff
@@ -4394,42 +4343,43 @@ An object that is equivalent to the date and time contained in `s`.
...
- All overloads are supported in: Windows 8  
+ [!code-csharp-interactive[System.DateTime.Parse#5](~/samples/snippets/csharp/VS_Snippets_CLR_System/system.DateTime.Parse/cs/Parse5.cs#5)]
+ [!code-vb[System.DateTime.Parse#5](~/samples/snippets/visualbasic/VS_Snippets_CLR_System/system.DateTime.Parse/vb/Parse5.vb#5)]  
   
+### Examples
+
+Numerous examples that call the `DateTime.Parse` method are interspersed throughout the [Remarks](#remarks) section of this article and in the documentation for the individual `DateTime.Parse` overloads.
+
+> [!NOTE]
+>  The C# examples in this article run in the [Try.NET](https://try.dot.net) inline code runner and playground. You can run an example in an interactive window by pressing the **Run** button. Once you execute the code, you can modify it and run the modified code by pressing **Run** again. The modified code either runs in the interactive window or, if compilation fails, the interactive window displays all C# compiler error messages.
```
Comment (pkulikov) [inline xml/System/DateTime.xml]: @rpetrusha there is an include file for this note: *~/includes/csharp-interactive-note.md*

**Review by rpetrusha — COMMENTED** [review]
Thanks for the careful review, @BillWagner. I've addressed all of your changes.

**Comment by rpetrusha** [comment]
I agree, @BillWagner, that we should have a single location for downloads. Although I think that we should combine DateTime and DateTime.Parse, I also think we should make them separately downloadable as well for those who just want the DateTime structure examples and those who just want the Parse examples.

**Review by BillWagner — APPROVED** [review]

**Comment by BillWagner** [comment]
@rpetrusha I looked at the updates, and this look great. It's much improved.

One item that should be checked when the build finishes: The table that was rendering incorrectly needs to be checked to make sure that's fixed. (The source looks good, but the is still running).

**Comment by rpetrusha** [comment]
Thanks, @BillWagner. The tables have all rendered correctly.
